File: benchmarks/padding_bench.cpp
#include "lfqueue/mpmc_queue.hpp"

#include <atomic>
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <new>
#include <stdexcept>
#include <thread>
#include <type_traits>
#include <utility>
#include <vector>

// ---------------------------------------------------------------------------
// Deliberately unpadded MPMC queue — head_ and tail_ share cache lines with
// buffer_ metadata, causing false sharing between producers and consumers.
// Kept here (not in the public header) because it is only useful for
// demonstrating the performance cost of missing padding.
// ---------------------------------------------------------------------------
template <typename T>
class NoPadMPMCQueue {
    static_assert(std::is_nothrow_move_constructible<T>::value, "");
    static_assert(std::is_nothrow_move_assignable<T>::value, "");

public:
    explicit NoPadMPMCQueue(std::size_t capacity)
        : capacity_(checked_pow2(capacity)), mask_(capacity_ - 1),
          buffer_(capacity_) {
        for (std::size_t i = 0; i < capacity_; ++i)
            buffer_[i].seq.store(i, std::memory_order_relaxed);
    }

    ~NoPadMPMCQueue() noexcept {
        std::size_t h = head_.load(std::memory_order_relaxed);
        std::size_t t = tail_.load(std::memory_order_relaxed);
        for (std::size_t p = h; p != t; ++p) {
            auto& slot = buffer_[p & mask_];
            if (slot.seq.load(std::memory_order_acquire) == p + 1)
                slot.data()->~T();
        }
    }

    NoPadMPMCQueue(const NoPadMPMCQueue&) = delete;
    NoPadMPMCQueue& operator=(const NoPadMPMCQueue&) = delete;

    bool push(T value) {
        std::size_t pos = tail_.load(std::memory_order_relaxed);
        for (;;) {
            auto& slot = buffer_[pos & mask_];
            std::size_t seq = slot.seq.load(std::memory_order_acquire);
            auto diff = static_cast<std::intptr_t>(seq) -
                        static_cast<std::intptr_t>(pos);
            if (diff == 0) {
                if (tail_.compare_exchange_weak(pos, pos + 1,
                        std::memory_order_relaxed, std::memory_order_relaxed)) {
                    slot.construct(std::move(value));
                    slot.seq.store(pos + 1, std::memory_order_release);
                    return true;
                }
            } else if (diff < 0) {
                return false;
            } else {
                pos = tail_.load(std::memory_order_relaxed);
            }
        }
    }

    bool pop(T& out) {
        std::size_t pos = head_.load(std::memory_order_relaxed);
        for (;;) {
            auto& slot = buffer_[pos & mask_];
            std::size_t seq = slot.seq.load(std::memory_order_acquire);
            auto diff = static_cast<std::intptr_t>(seq) -
                        static_cast<std::intptr_t>(pos + 1);
            if (diff == 0) {
                if (head_.compare_exchange_weak(pos, pos + 1,
                        std::memory_order_relaxed, std::memory_order_relaxed)) {
                    out = std::move(*slot.data());
                    slot.data()->~T();
                    slot.seq.store(pos + capacity_, std::memory_order_release);
                    return true;
                }
            } else if (diff < 0) {
                return false;
            } else {
                pos = head_.load(std::memory_order_relaxed);
            }
        }
    }

private:
    using Storage = typename std::aligned_storage<sizeof(T), alignof(T)>::type;

    struct Slot {
        std::atomic<std::size_t> seq{0};
        Storage storage;
        void construct(T&& v) { ::new (static_cast<void*>(&storage)) T(std::move(v)); }
        T* data() noexcept { return std::launder(reinterpret_cast<T*>(&storage)); }
    };

    static std::size_t checked_pow2(std::size_t n) {
        if (n == 0 || (n & (n - 1)) != 0)
            throw std::invalid_argument("NoPadMPMCQueue capacity must be a non-zero power of two");
        return n;
    }

    const std::size_t capacity_;
    const std::size_t mask_;
    std::vector<Slot> buffer_;
    // NOT cache-line padded — sits adjacent to buffer_ metadata,
    // and head_/tail_ share a cache line with each other.
    std::atomic<std::size_t> head_{0};
    std::atomic<std::size_t> tail_{0};
};

// ---------------------------------------------------------------------------
// Benchmark driver
// ---------------------------------------------------------------------------
template <typename Queue>
double run_throughput(int P, int C, long ops_per_producer) {
    Queue queue(1024);
    const long total = static_cast<long>(P) * ops_per_producer;
    std::atomic<long> consumed{0};
    std::atomic<int>  ready{0};
    std::atomic<bool> go{false};

    std::vector<std::thread> threads;
    threads.reserve(static_cast<std::size_t>(P + C));

    for (int p = 0; p < P; ++p) {
        threads.emplace_back([&]() {
            ready.fetch_add(1, std::memory_order_relaxed);
            while (!go.load(std::memory_order_acquire)) {}
            for (long i = 0; i < ops_per_producer; ++i)
                while (!queue.push(0)) {}
        });
    }
    for (int c = 0; c < C; ++c) {
        threads.emplace_back([&]() {
            ready.fetch_add(1, std::memory_order_relaxed);
            while (!go.load(std::memory_order_acquire)) {}
            int val;
            while (consumed.load(std::memory_order_relaxed) < total)
                if (queue.pop(val))
                    consumed.fetch_add(1, std::memory_order_relaxed);
        });
    }

    while (ready.load(std::memory_order_acquire) < P + C) {}
    auto t0 = std::chrono::steady_clock::now();
    go.store(true, std::memory_order_release);
    for (auto& t : threads) t.join();
    auto t1 = std::chrono::steady_clock::now();

    double seconds = std::chrono::duration<double>(t1 - t0).count();
    return static_cast<double>(total) / seconds / 1e6;
}

int main() {
    constexpr int P = 4, C = 4;
    constexpr long kTotal = 10'000'000;
    long ops = kTotal / P;
    long total = static_cast<long>(P) * ops;

    double nopad_mops = run_throughput<NoPadMPMCQueue<int>>(P, C, ops);
    double pad_mops   = run_throughput<lfqueue::MPMCQueue<int>>(P, C, ops);

    std::puts("configuration,total_ops,seconds,mops_per_sec");
    std::printf("without_padding,%ld,%.3f,%.1f\n",
                total, total / (nopad_mops * 1e6), nopad_mops);
    std::printf("with_padding,%ld,%.3f,%.1f\n",
                total, total / (pad_mops * 1e6), pad_mops);
    return 0;
}

File: benchmarks/padding_bench.rs
use oxide_broker::MpmcQueue;
use std::cell::UnsafeCell;
use std::mem::MaybeUninit;
use std::sync::atomic::{AtomicBool, AtomicI64, AtomicUsize, Ordering};
use std::thread;
use std::time::Instant;

// ---------------------------------------------------------------------------
// Deliberately unpadded MPMC queue — head and tail share cache lines with
// each other and with buffer metadata, causing false sharing between
// producers and consumers. Kept here (not in the library) because it is
// only useful for demonstrating the performance cost of missing padding.
// ---------------------------------------------------------------------------
struct Slot<T> {
    sequence: AtomicUsize,
    data: UnsafeCell<MaybeUninit<T>>,
}

struct NoPadMpmcQueue<T> {
    capacity: usize,
    mask: usize,
    buffer: Box<[Slot<T>]>,
    // NOT cache-line padded — head and tail sit adjacent to each other and
    // to buffer's metadata.
    head: AtomicUsize,
    tail: AtomicUsize,
}

unsafe impl<T: Send> Send for NoPadMpmcQueue<T> {}
unsafe impl<T: Send> Sync for NoPadMpmcQueue<T> {}

impl<T> NoPadMpmcQueue<T> {
    fn new(capacity: usize) -> Self {
        assert!(capacity != 0 && (capacity & (capacity - 1)) == 0);
        let buffer: Box<[Slot<T>]> = (0..capacity)
            .map(|i| Slot {
                sequence: AtomicUsize::new(i),
                data: UnsafeCell::new(MaybeUninit::uninit()),
            })
            .collect::<Vec<_>>()
            .into_boxed_slice();

        Self {
            capacity,
            mask: capacity - 1,
            buffer,
            head: AtomicUsize::new(0),
            tail: AtomicUsize::new(0),
        }
    }

    fn push(&self, value: T) -> Result<(), T> {
        let mut pos = self.tail.load(Ordering::Relaxed);
        loop {
            let slot = &self.buffer[pos & self.mask];
            let seq = slot.sequence.load(Ordering::Acquire);
            let diff = seq as isize - pos as isize;
            if diff == 0 {
                if self
                    .tail
                    .compare_exchange_weak(pos, pos + 1, Ordering::Relaxed, Ordering::Relaxed)
                    .is_ok()
                {
                    unsafe {
                        (*slot.data.get()).write(value);
                    }
                    slot.sequence.store(pos + 1, Ordering::Release);
                    return Ok(());
                }
            } else if diff < 0 {
                return Err(value);
            } else {
                pos = self.tail.load(Ordering::Relaxed);
            }
        }
    }

    fn pop(&self) -> Option<T> {
        let mut pos = self.head.load(Ordering::Relaxed);
        loop {
            let slot = &self.buffer[pos & self.mask];
            let seq = slot.sequence.load(Ordering::Acquire);
            let diff = seq as isize - (pos + 1) as isize;
            if diff == 0 {
                if self
                    .head
                    .compare_exchange_weak(pos, pos + 1, Ordering::Relaxed, Ordering::Relaxed)
                    .is_ok()
                {
                    let value = unsafe { (*slot.data.get()).assume_init_read() };
                    slot.sequence.store(pos + self.capacity, Ordering::Release);
                    return Some(value);
                }
            } else if diff < 0 {
                return None;
            } else {
                pos = self.head.load(Ordering::Relaxed);
            }
        }
    }
}

impl<T> Drop for NoPadMpmcQueue<T> {
    fn drop(&mut self) {
        let head = self.head.load(Ordering::Relaxed);
        let tail = self.tail.load(Ordering::Relaxed);
        let mut pos = head;
        while pos != tail {
            let slot = &self.buffer[pos & self.mask];
            if slot.sequence.load(Ordering::Relaxed) == pos + 1 {
                unsafe {
                    (*slot.data.get()).assume_init_drop();
                }
            }
            pos += 1;
        }
    }
}

// ---------------------------------------------------------------------------
// Benchmark driver
// ---------------------------------------------------------------------------
trait Pushable<T> {
    fn try_push(&self, value: T) -> Result<(), T>;
    fn try_pop(&self) -> Option<T>;
}

impl<T> Pushable<T> for NoPadMpmcQueue<T> {
    fn try_push(&self, value: T) -> Result<(), T> {
        self.push(value)
    }
    fn try_pop(&self) -> Option<T> {
        self.pop()
    }
}

impl<T> Pushable<T> for MpmcQueue<T> {
    fn try_push(&self, value: T) -> Result<(), T> {
        self.push(value)
    }
    fn try_pop(&self) -> Option<T> {
        self.pop()
    }
}

fn run_throughput<Q>(queue: Q, producers: usize, consumers: usize, ops_per_producer: i64) -> f64
where
    Q: Pushable<i32> + Send + Sync,
{
    let total = producers as i64 * ops_per_producer;
    let consumed = AtomicI64::new(0);
    let ready = AtomicUsize::new(0);
    let go = AtomicBool::new(false);

    let start = thread::scope(|scope| {
        for _ in 0..producers {
            let queue = &queue;
            let ready = &ready;
            let go = &go;
            scope.spawn(move || {
                ready.fetch_add(1, Ordering::Relaxed);
                while !go.load(Ordering::Acquire) {
                    std::hint::spin_loop();
                }
                for _ in 0..ops_per_producer {
                    while queue.try_push(0).is_err() {}
                }
            });
        }
        for _ in 0..consumers {
            let queue = &queue;
            let ready = &ready;
            let go = &go;
            let consumed = &consumed;
            scope.spawn(move || {
                ready.fetch_add(1, Ordering::Relaxed);
                while !go.load(Ordering::Acquire) {
                    std::hint::spin_loop();
                }
                while consumed.load(Ordering::Relaxed) < total {
                    if queue.try_pop().is_some() {
                        consumed.fetch_add(1, Ordering::Relaxed);
                    }
                }
            });
        }

        while ready.load(Ordering::Acquire) < (producers + consumers) {
            std::hint::spin_loop();
        }
        let start = Instant::now();
        go.store(true, Ordering::Release);
        start
    });

    let seconds = start.elapsed().as_secs_f64();
    total as f64 / seconds / 1e6
}

fn main() {
    const P: usize = 4;
    const C: usize = 4;
    const TOTAL: i64 = 10_000_000;
    let ops = TOTAL / P as i64;
    let total = P as i64 * ops;

    let nopad_mops = run_throughput(NoPadMpmcQueue::<i32>::new(1024), P, C, ops);
    let pad_mops = run_throughput(MpmcQueue::<i32>::new(1024), P, C, ops);

    println!("configuration,total_ops,seconds,mops_per_sec");
    println!(
        "without_padding,{},{:.3},{:.1}",
        total,
        total as f64 / (nopad_mops * 1e6),
        nopad_mops
    );
    println!(
        "with_padding,{},{:.3},{:.1}",
        total,
        total as f64 / (pad_mops * 1e6),
        pad_mops
    );
}
